File: clone.zig
const std = @import("std");
const linux = std.os.linux;

fn child(arg: usize) callconv(.c) u8 {
    std.debug.print("child: pid: '{d}', arg: {d}\n", .{ linux.getpid(), arg });
    return 0;
}

pub fn main() !void {
    const allocator = std.heap.page_allocator;
    const stack_size = 1024 * 1024;
    const stack = try allocator.alloc(u8, stack_size);

    const flags: u32 = linux.CLONE.VM;

    const stack_ptr = @intFromPtr(stack.ptr + stack_size);
    const child_pid = linux.clone(
        &child,
        stack_ptr,
        flags,
        0,
        null,
        0,
        null,
    );

    var status: u32 = undefined;
    const waitpid_flags: u32 = 0;
    const rc = linux.waitpid(@intCast(child_pid), &status, waitpid_flags);

    std.debug.print("parent: child exited with status: '{d}', pid: '{d}', wait_pid: '{d}'\n", .{
        child_pid,
        linux.getpid(),
        rc,
    });
}

File: clone.c
#define _GNU_SOURCE
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sched.h>


static int child(void *arg) {
	char *input = (char *)arg;

	printf("child:args = '%s', tid = '%d', pid = '%d'\n", input, gettid(), getpid());
	strcpy(input, "arguments received successfully");
	return 0;
}

int main(int argc, char **argv) {
	
	// allocate stack for the child
	const int STACK_SIZE = 1024 * 1024;
	char *stack = malloc(STACK_SIZE);
	if (!stack) {
		perror("malloc");
		exit(1);
	}

	unsigned long flags = 0;
	if (argc > 1 && !strcmp(argv[1], "vm")) {
		flags |= CLONE_VM;
	}

	char buf[100];
	strcpy(buf, "gcc -wall -ansi -Wall -pedantic");

	if (clone(child, stack + STACK_SIZE, flags | SIGCHLD, &buf) == -1) {
		perror("clone");
		exit(1);
	}

	int status;
	if(wait(&status) == -1) {
		perror("wait");
		exit(1);
	}

	printf("parenet: child exited with status '%d', tid = '%d', pid = '%d', message from child = '%s'\n", status, gettid(), getpid(), buf);
	return 0;
}
